// safe-routes@1.0.18/dist/vite.js
import { reactRouter } from "@react-router/dev/vite";
import replace from '@rollup/plugin-replace';
import { join } from 'node:path';
import { placeholder } from './basename.js';
import { DEFAULT_OUTPUT_DIR_PATH, TYPE_FILE_NAME, build } from './build.js';
function extractReactRouterPluginContext(config) {
    const reactRouterPluginContextName = '__reactRouterPluginContext';
    if (reactRouterPluginContextName in config) {
        return config[reactRouterPluginContextName];
    }
    return null;
}
function findReactRouterPlugin(plugins) {
    return plugins.find((plugin) => plugin.name === 'react-router');
}
export function safeRoutes(pluginConfig = {}) {
    let reactRouterPlugin;
    let rootDirectory;
    let viteUserConfig;
    let viteConfigEnv;
    let ctx;
    function generateTypeFile() {
        if (!ctx) {
            return;
        }
        build(rootDirectory, ctx.reactRouterConfig, { strict: pluginConfig.strict, outputDirPath: pluginConfig.outDir || DEFAULT_OUTPUT_DIR_PATH });
    }
    async function reloadCtx() {
        const plugin = findReactRouterPlugin(reactRouter());
        if (!plugin || !plugin.config) {
            return;
        }
        const config = await plugin.config(viteUserConfig, { ...viteConfigEnv, command: 'build' });
        ctx = extractReactRouterPluginContext(config);
    }
    return {
        name: 'safe-routes',
        enforce: 'post',
        config(_viteUserConfig, _viteConfigEnv) {
            var _a;
            viteUserConfig = _viteUserConfig;
            viteConfigEnv = _viteConfigEnv;
            if (ctx && ctx.reactRouterConfig.basename) {
                (_a = viteUserConfig.plugins) === null || _a === void 0 ? void 0 : _a.push(replace.default({
                    [placeholder]: ctx.reactRouterConfig.basename,
                }));
            }
        },
        configResolved(config) {
            reactRouterPlugin = findReactRouterPlugin(config.plugins);
            if (!reactRouterPlugin) {
                return;
            }
            rootDirectory = config.root;
            ctx = extractReactRouterPluginContext(config);
        },
        async configureServer() {
            generateTypeFile();
        },
        async watchChange(id) {
            if (!reactRouterPlugin || !ctx) {
                return;
            }
            if (!id.startsWith(ctx.reactRouterConfig.appDirectory)) {
                return;
            }
            if (pluginConfig.outDir && id === join(rootDirectory, pluginConfig.outDir, TYPE_FILE_NAME)) {
                return;
            }
            await reloadCtx();
            generateTypeFile();
        },
    };
}
